File: src/main/python/g0101_0200/s0124_binary_tree_maximum_path_sum/Solution.py
# #Hard #Top_100_Liked_Questions #Top_Interview_Questions #Dynamic_Programming #Depth_First_Search
# #Tree #Binary_Tree #Udemy_Tree_Stack_Queue #Top_Interview_150_Binary_Tree_General
# #Big_O_Time_O(N)_Space_O(N) #2024_06_09_Time_57_ms_(86.71%)_Space_20.9_MB_(40.96%)

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def __init__(self):
        self.max_sum = float('-inf')

    def maxPathSum(self, root: Optional[TreeNode]) -> int:
        self.helper(root)
        return self.max_sum

    def helper(self, root):
        if not root:
            return 0
        
        left = max(0, self.helper(root.left))
        right = max(0, self.helper(root.right))
        
        current = root.val + left + right
        self.max_sum = max(self.max_sum, current)
        
        return root.val + max(left, right)

File: src/main/swift/g0101_0200/s0124_binary_tree_maximum_path_sum/Solution.swift
// #Hard #Top_100_Liked_Questions #Top_Interview_Questions #Dynamic_Programming #Depth_First_Search
// #Tree #Binary_Tree #Udemy_Tree_Stack_Queue #Top_Interview_150_Binary_Tree_General
// #Big_O_Time_O(N)_Space_O(N) #2024_06_27_Time_28_ms_(98.45%)_Space_17.1_MB_(67.36%)

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public var val: Int
 *     public var left: TreeNode?
 *     public var right: TreeNode?
 *     public init() { self.val = 0; self.left = nil; self.right = nil; }
 *     public init(_ val: Int) { self.val = val; self.left = nil; self.right = nil; }
 *     public init(_ val: Int, _ left: TreeNode?, _ right: TreeNode?) {
 *         self.val = val
 *         self.left = left
 *         self.right = right
 *     }
 * }
 */
public class Solution {
    private var maxSum = Int.min

    private func helper(_ root: TreeNode?) -> Int {
        guard let root = root else {
            return 0
        }

        let left = max(0, helper(root.left))
        let right = max(0, helper(root.right))
        let current = root.val + left + right
        maxSum = max(maxSum, current)

        return root.val + max(left, right)
    }

    public func maxPathSum(_ root: TreeNode?) -> Int {
        _ = helper(root)
        return maxSum
    }
}
